src/core/cluster.py

```diff
@@ -26,18 +28,22 @@
     SECRETS_UNIT,
     SECURITY_PROTOCOL_PORTS,
     ZK,
-    AuthMechanism,
+    AuthMap,
     Status,
     Substrates,
 )
 
+if TYPE_CHECKING:
+    from charm import KafkaCharm
+
 
 class ClusterState(Object):
     """Collection of global cluster state for the Kafka services."""
 
-    def __init__(self, charm: Framework | Object, substrate: Substrates):
+    def __init__(self, charm: "KafkaCharm", substrate: Substrates):
         super().__init__(parent=charm, key="charm_state")
         self.substrate: Substrates = substrate
+        self.config = charm.config
```

Comment:
**huge suggestion** If this is the reason for having the charm in the constructor, why not passing as argument the `config` object? The definition of the config object is still in `core` (in structure_config), so we would not need to break the clean import (here we are importing from charm which is at the end of the dependency tree) for the typing (yes, I understand it is only for the typing, but we really don't need this I believe)

Reply:
Lol, I had originally done it that way, see https://github.com/canonical/kafka-k8s-operator/pull/110#discussion_r1696552026

@zmraul - what was your reasoning here?

Reply:
My 2 cents would be to pass the necessary and sufficient dependencies, so, like the original work did. The reasoning would usually be around surface attack, dependency tree, and ease to test stuff (for instance, we would not have to instantiate a complete `KafkaCharm` to run unit tests on `ClusterState`.
That being said, we do not have those type of laser focused unit tests, and it can be convenient to pass `charm` rather than X different attributes.

tl;dr: really tiny +0.01 on passing `config`

Reply:
To me it is for both keeping scope confined AND clear dependency tree. It is definitely more than 0.01 for me :joy:

Reply:
Keeping the typing as `Framework | Object` is fine. I think that if we find more cases like this, where`cluster` or `core` are in need of accessing things that come from charm in execution time, we are forcing the dependency tree into what we _actually_ need. 
Passing as an explicit argument is a good way to keep scope and dependency, but to me it points to the code not being able to strictly follow the pattern.

Reply:
While I'm fine with either option, I will say that the requirement to avoid passing 'charm' around **was** supposed to make mocking state/config easier with Scenario. For those that have used Scenario, was it any easier than with `harness.charm`?

Reply:
Tiny nitpick, `Framework` is a subclass of `Object` so there is probably no need for the union of types. I do not remember any control flow related to that.

> While I'm fine with either option, I will say that the requirement to avoid passing 'charm' around was supposed to make mocking state/config easier with Scenario. For those that have used Scenario, was it any easier than with harness.charm?

Hum, I would say that there is no difference at all, because Scenario entry point [is the charm itself](https://github.com/canonical/ops-scenario/blob/main/scenario/context.py#L234), you don't test an event handler in isolation. Though I could be wrong, but the more I think about it, the more I believe that given sufficient complexity, scenario would work best with simple test charms dedicated to testing one event handler at a time